URL unsafe characters get decoded before the request is sent

If using the Apache Http Client, when the path contains URL encoded unsafe characters, they get decoded while setting up the RestTemplate, causing an internal exception.

### Feign config

```java
@Configuration
public class FeignConfig {

    @Bean
    public TestClient testClient() {
        return Feign.builder()
                .client(new ApacheHttpClient(org.apache.http.impl.client.HttpClientBuilder.create().build()))
                .target(TestClient.class, "https://website.com/path");
    }
}
```

### Client

```java
public interface TestClient {
    @RequestLine("GET /")
    String findSomething(URI uri);
}
```

### Usage

```java
@Component
public class TestComponent {

    private final TestClient testClient;

    public TestComponent(TestClient testClient) {
        this.testClient = testClient;
    }

    @PostConstruct
    public void test() {
        testClient.findSomething(URI.create("https://website.com/path/text%7Ctext"));
    }
}
```

### Result

```
Caused by: feign.RetryableException: URL 'https://website.com/path/text|text/' couldn't be parsed into a URI executing GET https://website.com/path/text|text/
	at feign.FeignException.errorExecuting(FeignException.java:277) ~[feign-core-12.4.jar:na]
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:110) ~[feign-core-12.4.jar:na]
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:70) ~[feign-core-12.4.jar:na]
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:96) ~[feign-core-12.4.jar:na]
	at jdk.proxy2/jdk.proxy2.$Proxy43.findSomething(Unknown Source) ~[na:na]
	at com.example.test.TestComponent.test(TestComponent.java:18) ~[classes/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:577) ~[na:na]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMethod.invoke(InitDestroyAnnotationBeanPostProcessor.java:457) ~[spring-beans-6.0.13.jar:6.0.13]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:401) ~[spring-beans-6.0.13.jar:6.0.13]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:219) ~[spring-beans-6.0.13.jar:6.0.13]
	... 17 common frames omitted
Caused by: java.io.IOException: URL 'https://website.com/path/text|text/' couldn't be parsed into a URI
	at feign.httpclient.ApacheHttpClient.execute(ApacheHttpClient.java:79) ~[feign-httpclient-12.4.jar:na]
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:100) ~[feign-core-12.4.jar:na]
	... 26 common frames omitted
Caused by: java.net.URISyntaxException: Illegal character in path at index 29: https://website.com/path/text|text/
	at java.base/java.net.URI$Parser.fail(URI.java:2973) ~[na:na]
	at java.base/java.net.URI$Parser.checkChars(URI.java:3144) ~[na:na]
	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3226) ~[na:na]
	at java.base/java.net.URI$Parser.parse(URI.java:3174) ~[na:na]
	at java.base/java.net.URI.<init>(URI.java:623) ~[na:na]
	at org.apache.http.client.utils.URIBuilder.<init>(URIBuilder.java:82) ~[httpclient-4.5.14.jar:4.5.14]
	at feign.httpclient.ApacheHttpClient.toHttpUriRequest(ApacheHttpClient.java:99) ~[feign-httpclient-12.4.jar:na]
	at feign.httpclient.ApacheHttpClient.execute(ApacheHttpClient.java:77) ~[feign-httpclient-12.4.jar:na]
	... 27 common frames omitted
```

--- a/core/src/main/java/feign/RequestTemplate.java
+++ b/core/src/main/java/feign/RequestTemplate.java
@@ -518,7 +518,7 @@ public final class RequestTemplate implements Serializable {
       }
 
       /* strip the query string */
-      this.target = targetUri.getScheme() + "://" + targetUri.getAuthority() + targetUri.getPath();
+      this.target = targetUri.getScheme() + "://" + targetUri.getRawAuthority() + targetUri.getRawPath();
       if (targetUri.getFragment() != null) {
         this.fragment = "#" + targetUri.getFragment();
       }

--- a/core/src/test/java/feign/RequestTemplateTest.java
+++ b/core/src/test/java/feign/RequestTemplateTest.java
@@ -515,6 +515,15 @@ public class RequestTemplateTest {
   }
 
   @Test
+  public void urlEncodingRemainsInPlace() {
+    RequestTemplate template = new RequestTemplate()
+        .method(HttpMethod.GET)
+        .target("https://exa%23mple.com/path%7Cpath");
+
+    assertThat(template.url()).isEqualTo("https://exa%23mple.com/path%7Cpath");
+  }
+
+  @Test
   public void slashShouldNotBeAppendedForMatrixParams() {
     RequestTemplate template = new RequestTemplate().method(HttpMethod.GET)
         .uri("/path;key1=value1;key2=value2", true);

